### ISSUE TITLE
Document Update Request: Update example to build custom Che binaries for Che version 6

### ISSUE THREAD

**Issue by VinHou** [eclipse-che/che#8219]
**Description**

We are following this documentation link https://www.eclipse.org/che/docs/assemblies/assembly-lifecycle/index.html#production-mode to build our custom Che binaries in a production mode. The docker file structure as specified in step 2 links us to https://github.com/codenvy/plugin-openshift/tree/master/dockerfiles. This example directory structure is not updated for version 6.

Please provide us with an updated example for version 6 so that we can migrate our custom Che binaries from Che 5.17.0 to Che 6.0.0.

**Comment by VinHou** [comment]
We got the following error when running Che by using our custom Docker images migrated from 5.17.0 to Che 6.0.0:

cp: can't stat '/home/user/eclipse-che//lib/*': No such file or directory

**Comment by ghost** [comment]
@VinHou can you share more details on what exactly you have done? You have built your own CLI, right?

**Comment by ghost** [comment]
@VinHou btw, if you have your own Che server image, you can use existing CLI, just pass `-e "IMAGE_CHE=yourImage"`

And yes, I will review docs of course.

**Comment by VinHou** [comment]
@eivantsov We are able to build our own design-server/dts-cli, design-server/dts-init, and design-server/dts-server when we migrated Che from 5.17.0 to 6.0.0-M4. But when we tried to start Che by using our design-server/dts-cli, we got the following error:

cp: can't stat '/home/user/eclipse-che//lib/*': No such file or directory

We don't have any problem when using Che 5.17.0 before.

**Comment by VinHou** [comment]
@eivantsov I tried the following command by passing -e "IMAGE_CHE=design-server/dts-cli:11.0.0" as you suggested:

sudo docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock \
           -e "IMAGE_CHE=design-server/dts-cli:11.0.0" \
           -v /home/rasp/che_data:/data \
            design-server/dts-cli:11.0.0 start \
            --skip:scripts

I still could not start che and the “sudo docker logs -f che” command shows the following usage error:

USAGE:
  docker run -it --rm <DOCKER_PARAMETERS> design-server/dts-cli:11.0.0 [COMMAND]

It appears that IMAGE_CHE is an invalid environment variable. I then used the following command:

sudo docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock \
           -e "CHE_IMAGE=design-server/dts-cli:11.0.0" \
           -v /home/rasp/che_data:/data \
            design-server/dts-cli:11.0.0 start \
            --skip:scripts

Namely, changed IMAGE_CHE to CHE_IMAGE. This time I didn’t get the usage error but got the same error we had before:

cp: can't stat '/home/user/eclipse-che//lib/*': No such file or directory

It seems that passing an environment variable for CHE_IMAGE is a redundant one.

**Comment by ghost** [comment]
@VinHou this is exactly how I start custom binaries `-e IMAGE_CHE=`, though I use eclipse/che:nightly as a CLI image. You have got a custom one.

In your case you have passed an env that does not mean anything to CLI, so I wonder how this resolved your problem.

You may use own init image too `-e IMAGE_INIT` but use eclipse/che:6.0.0-M4 cli image.

I have just ran `docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock -e "IMAGE_CHE=eivantsov/che:js" -v ~/Documents/chejss:/data eclipse/che:nightly start` and it worked perfectly well for me - I can see that it's my custom binaries are running there.

**Comment by VinHou** [comment]
@eivantsov I tried the following command:

sudo docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock \
           -e "IMAGE_INIT=design-server/dts-init:11.0.0" \
           -v /home/rasp/che_data:/data \
            design-server/dts-cli:11.0.0 start \
            --skip:scripts

But still got the same error:

cp: can't stat '/home/user/eclipse-che//lib/*': No such file or directory

Any other suggestion?

**Comment by VinHou** [comment]
@eivantsov  I also tried the following command:

sudo docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock \
           -e "IMAGE_CHE=design-server/dts-server:11.0.0" \
           -v /home/rasp/che_data:/data \
            design-server/dts-cli:11.0.0 start \
            --skip:scripts

This time I didn't get any usage error but got the same error:

cp: can't stat '/home/user/eclipse-che//lib/*': No such file or directory

**Comment by ghost** [comment]
@VinHou since your image is private and I cannot peek into it, here's smth that may help you find out what's going wrong - https://github.com/eclipse/che/blob/93b974d1e0ec7836635f3d7da196b7651da8ad1b/dockerfiles/che/entrypoint.sh#L254

This is the entrypoint that gets executed when Che server container starts. If I had the same issue I'd go for:

```
docker run -ti --entrypoint="/bin/bash" design-server/dts-server:11.0.0
cd ~/eclipse/che
ls -la
```

Is there a lib folder there? If not, then your custom binaries weren't built the right way.

**Comment by VinHou** [comment]
@eivantsov Many thanks for your information. I tried the entrypoint test as suggested and got the following result:

rasp@usrscp30:~/softwareag-che/dockerfiles/che$ sudo docker run -ti --entrypoint="/bin/bash" design-server/dts-server:11.0.0
[sudo] password for rasp:
bash-4.3# cd ~/eclipse/che
bash: cd: /root/eclipse/che: No such file or directory

We did think that our custom binaries might not be built correctly for Che version 6 at the beginning. That's the reason why we requested an updated example to build custom Che binaries for Che version 6.

**Comment by ghost** [comment]
I'm sorry, it should be `cd /home/user/eclipse-che`

**Comment by VinHou** [comment]
@eivantsov I tried cd /home/user/eclipse-che and got the same result.

rasp@usrscp30:~/softwareag-che/dockerfiles/che$ sudo docker run -ti --entrypoint="/bin/bash" design-server/dts-server:11.0.0
[sudo] password for rasp:
bash-4.3# cd /home/user/eclipse-che
bash: cd: /home/user/eclipse-che: No such file or directory
bash-4.3#

**Comment by ghost** [comment]
@VinHou there are no binaries in your image then.

Did you use this script? https://github.com/eclipse/che/blob/master/dockerfiles/che/build.sh?

Just copy the entire dockerfiles dir to your custom assembly and execute this run.sh

**Comment by VinHou** [comment]
@eivantsov We have our own build.sh similar to your https://github.com/eclipse/che/blob/master/dockerfiles/che/build.sh and we have our own Dockerfile similar to your https://github.com/eclipse/che/blob/master/dockerfiles/che/Dockerfile to build our custom binaries.

Actually, we use /home/user/eclipse-che-11.0.0 instead of /home/user/eclipse-che. Namely, we added our DTS version to the end. I executed entrypoint again and we do have /home/user/eclipse-che-11.0.0 as shown below:

rasp@usrscp30:~/softwareag-che/dockerfiles/che$ sudo docker run -ti --entrypoint="/bin/bash" design-server/dts-server:11.0.0
[sudo] password for rasp:
bash-4.3# cd /home/user/eclipse-che-11.0.0
bash-4.3# pwd
/home/user/eclipse-che-11.0.0
bash-4.3# ls
lib        templates  tomcat
bash-4.3#

**Comment by ghost** [comment]
@VinHou since I cannot see your scripts and Dockerfiles, the only thing I can suggest is pass `-e CHE_HOME=/home/user/eclipse-che-11.0.0` which should do the trick.

**Comment by VinHou** [comment]
@eivantsov I tried the following command by adding CHE_HOME as suggested:

sudo docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock \
           -e CHE_HOME=/home/user/eclipse-che-11.0.0 \
           -v /home/rasp/che_data:/data \
            design-server/dts-cli:11.0.0 start \
            --skip:scripts

But still got the same error:
 
cp: can't stat '/home/user/eclipse-che//lib/*': No such file or directory

**Comment by ghost** [comment]
@VinHou I cannot tell you what to do since I have no access to your dockerfiles and scripts. If you build your own che-server image, why don't you modify https://github.com/eclipse/che/blob/93b974d1e0ec7836635f3d7da196b7651da8ad1b/dockerfiles/che/entrypoint.sh#L211 yourself? Or rebuild your che-server image so that your folder is named the way entrypoint expects it?

**Comment by VinHou** [comment]
@eivantsov Please keep us informed if https://www.eclipse.org/che/docs/assemblies/assembly-lifecycle/index.html#production-mode is updated for Che 6. We need to use an updated example to build our custom Che binaries.

**Comment by stnnguyen95** [comment]
@eivantsow. This is a follow-up of the work previously done by @VinHou. I got the source of Che 6 and copied from the che/dockerfiles folder the minimum needed to build, namely the base, bats, che, cli and init folders. I modified a few properties as described in https://www.eclipse.org/che/docs/assemblies/assembly-lifecycle/index.html#production-mode. Since there are major changes between Che 5 and Che 6, I did not try to match the base Che images.
The build was successful. I was able to start Che and can see our custom plugins and extensions.
Now, the question is how do I set the base Che images for the new scripts from Che 6 source. This was well documented for Che 6 as already mentioned here https://www.eclipse.org/che/docs/assemblies/assembly-lifecycle/index.html#production-mode.
